IOPool/Input/src/RootTree.cc

```diff
@@ -26,60 +24,75 @@ namespace edm {
                      BranchType const& branchType,
                      unsigned int nIndexes,
-                     unsigned int maxVirtualSize,
-                     unsigned int cacheSize,
                      unsigned int learningEntries,
                      bool enablePrefetching,
                      InputType inputType)
       : filePtr_(filePtr),
-        tree_(dynamic_cast<TTree*>(
-            filePtr_.get() != nullptr ? filePtr_->Get(BranchTypeToProductTreeName(branchType).c_str()) : nullptr)),
-        metaTree_(dynamic_cast<TTree*>(
-            filePtr_.get() != nullptr ? filePtr_->Get(BranchTypeToMetaDataTreeName(branchType).c_str()) : nullptr)),
         branchType_(branchType),
-        auxBranch_(tree_ ? getAuxiliaryBranch(tree_, branchType_) : nullptr),
-        treeCache_(),
-        rawTreeCache_(),
-        triggerTreeCache_(),
-        rawTriggerTreeCache_(),
-        trainedSet_(),
-        triggerSet_(),
-        entries_(tree_ ? tree_->GetEntries() : 0),
-        entryNumber_(-1),
-        entryNumberForIndex_(new std::vector<EntryNumber>(nIndexes, IndexIntoFile::invalidEntry)),
-        branchNames_(),
-        branches_{},
-        trainNow_(false),
-        switchOverEntry_(-1),
-        rawTriggerSwitchOverEntry_(-1),
-        performedSwitchOver_{false},
+        entryNumberForIndex_(std::make_unique<std::vector<EntryNumber>>(nIndexes, IndexIntoFile::invalidEntry)),
         learningEntries_(learningEntries),
-        cacheSize_(cacheSize),
-        treeAutoFlush_(0),
         enablePrefetching_(enablePrefetching),
         enableTriggerCache_(branchType_ == InEvent),
-        rootDelayedReader_(new RootDelayedReader(*this, filePtr, inputType)),
-        branchEntryInfoBranch_(metaTree_ ? getProductProvenanceBranch(metaTree_, branchType_)
-                                         : (tree_ ? getProductProvenanceBranch(tree_, branchType_) : nullptr)),
-        infoTree_(dynamic_cast<TTree*>(filePtr_.get() != nullptr
-                                           ? filePtr->Get(BranchTypeToInfoTreeName(branchType).c_str())
-                                           : nullptr))  // backward compatibility
-  {
+        rootDelayedReader_(std::make_unique<RootDelayedReader>(*this, filePtr, inputType)) {}
+
+  // Used for Event/Lumi/Run RootTrees
+  RootTree::RootTree(std::shared_ptr<InputFile> filePtr,
+                     BranchType const& branchType,
+                     unsigned int nIndexes,
+                     unsigned int maxVirtualSize,
+                     unsigned int cacheSize,
+                     unsigned int learningEntries,
+                     bool enablePrefetching,
+                     InputType inputType)
+      : RootTree(filePtr, branchType, nIndexes, learningEntries, enablePrefetching, inputType) {
+    init(BranchTypeToProductTreeName(branchType), maxVirtualSize, cacheSize);
+    metaTree_ = dynamic_cast<TTree*>(filePtr_->Get(BranchTypeToMetaDataTreeName(branchType).c_str()));
```

Comment:
filePtr_ and tree_ cannot be null. There are two things protecting this.

The code below gets executed before anything ever dereferences filePtr_ or tree_ in RootTree.cc. If you think it is worth it I could add an explicit check on filePtr_ and a different exception.
```
  void RootTree::init(std::string const& productTreeName, unsigned int maxVirtualSize, unsigned int cacheSize) {
    if (filePtr_.get() != nullptr) {
      tree_ = dynamic_cast<TTree*>(filePtr_->Get(productTreeName.c_str()));
    }
    if (not tree_) {
      throw cms::Exception("WrongFileFormat")
          << "The ROOT file does not contain a TTree named " << productTreeName
          << "\n This is either not an edm ROOT file or is one that has been corrupted.";
    }
```

Also if you look up the stack in RootInputFileSequence.cc:
```
    if (filePtr) {
      size_t currentIndexIntoFile = fileIter_ - fileIterBegin_;
      rootFile_ = makeRootFile(filePtr);
```
and RepeatingCachedRootSource.cc 
```
  auto input =
      std::make_shared<InputFile>(physicalFileName.c_str(), "  Initiating request to open file ", InputType::Primary);
  rootFile_ = makeRootFile(
      logicalFileName, physicalFileName, 0 != nEventsToSkip, input, skipper, duplicateChecker, indexesIntoFiles);
```
filePtr and input get passed down from those two code fragments and cannot be null.

Reply:
> If you think it is worth it I could add an explicit check on filePtr_ and a different exception.

I don't, I was just wondering because the old code did those checks.